### PR TITLE
doc: correct analyzepsbt rpc doc

### PR DESCRIPTION
**Pull request by fanquake**
Even though `missing` is optional, all its field are also all optional.

`estimated_vsize` is optional (calculate alongside `estimated_feerate`).

Make `estimated_feerate` output numeric.

### CODE REVIEW

File: src/rpc/rawtransaction.cpp [hunk 1]
```diff
@@ -1862,21 +1862,21 @@ UniValue analyzepsbt(const JSONRPCRequest& request)
...
-                "        \"signatures\" : [          (array)\n"
+                "        \"signatures\" : [          (array, optional)\n"
                 "          \"keyid\"                 (string) Public key ID, hash160 of the public key, of a public key whose signature is missing\n"
                 "        ]\n"
-                "        \"redeemscript\" : \"hash\"   (string) Hash160 of the redeemScript that is missing\n"
-                "        \"witnessscript\" : \"hash\"  (string) SHA256 of the witnessScript that is missing\n"
+                "        \"redeemscript\" : \"hash\"   (string, optional) Hash160 of the redeemScript that is missing\n"
+                "        \"witnessscript\" : \"hash\"  (string, optional) SHA256 of the witnessScript that is missing\n"
                 "      }\n"
                 "      \"next\" : \"role\"           (string) Role of the next person that this input needs to go to\n"
```
Comment (fanquake) [inline src/rpc/rawtransaction.cpp]: Looks like `next` is also optional. i.e for:
`bitcoin-cli analysepsbt "cHNidP8BAJoCAAAAAljoeiG1ba8MI76OcHBFbDNvfLqlyHV5JPVFiHuyq911AAAAAAD/////g40EJ9DsZQpoqka7CwmK6kQiwHGyyng1Kgd5WdB86h0BAAAAAP////8CcKrwCAAAAAAWABTYXCtx0AYLCcmIauuBXlCZHdoSTQDh9QUAAAAAFgAUAK6pouXw+HaliN9VRuh0LR2HAI8AAAAAAAEAuwIAAAABqtc5MQGL0l+ErkALaISL4J23BurCrBgpi6vucatlb4sAAAAASEcwRAIgWPb8fGoz4bMVSNSByCbAFb0wE1qtQs1neQ2rZtKtJDsCIEoc7SYExnNbY5PltBaR3XiwDwxZQvufdRhW+qk4FX26Af7///8CgPD6AgAAAAAXqRQPuUY0IWlrgsgzryQceMF9295JNIfQ8gonAQAAABepFCnKdPigj4GZlCgYXJe12FLkBj9hh2UAAAABB9oARzBEAiB0AYrUGACXuHMyPAAVcgs2hMyBI4kQSOfbzZtVrWecmQIgc9Npt0Dj61Pc76M4I8gHBRTKVafdlUTxV8FnkTJhEYwBSDBFAiEA9hA4swjcHahlo0hSdG8BV3KTQgjG0kRUOTzZm98iF3cCIAVuZ1pnWm0KArhbFOXikHTYolqbV2C+ooFvZhkQoAbqAUdSIQKVg785rgpgl0etGZrd1jT6YQhVnWxc05tMIYPxq5bgfyEC2rYf9JoU22p9ArDNH7t4/EsYMStbTlTa5Nui+/71NtdSrgABASAAwusLAAAAABepFLf1+vQOPUClpFmx2zU18rcvqSHohwEHIyIAIIwjUxc3Q7WV37Sge3K6jkLjeX2nTof+fZ10l+OyAokDAQjaBABHMEQCIGLrelVhB6fHP0WsSrWh3d9vcHX7EnWWmn84Pv/3hLyyAiAMBdu3Rw2/LwhVfdNWxzJcHtMJE+mWzThAlF2xIijaXwFHMEQCIGX0W6WZi1mif/4ae+0BavHx+Q1Us6qPdFCqX1aiUQO9AiB/ckcDrR7blmgLKEtW1P/LiPf7dZ6rvgiqMPKbhROD0gFHUiEDCJ3BDHrG21T5EymvYXMz2ziM6tDCMfcjN50bmQMLAtwhAjrdkE89bc9Z3bkGsN7iNSm3/7ntUOXoYVGSaGAiHw5zUq4AIgIDqaTDf1mW06ol26xrVwrwZQOUSSlCRgs1R1Ptnuylh3EQ2QxqTwAAAIAAAACABAAAgAAiAgJ/Y5l1fS7/VaE2rQLGhLGDi2VW5fG2s0KCqUtrUAUQlhDZDGpPAAAAgAAAAIAFAACAAA=="
`

```
{
  "inputs": [
    {
      "has_utxo": true,
      "is_final": true
    },
    {
      "has_utxo": true,
      "is_final": true
    }
  ],
  "next": "extractor",
  "estimated_vsize": 463,
  "estimated_feerate": "0.00021598 BTC/kB",
  "fee": 0.00010000
}
```

Will update.

**Comment by achow101** [comment]
> `estimated_feerate` is a string.

Eh, it's not supposed to be.

**Comment by laanwj** [comment]
> `"estimated_feerate": "0.00021598 BTC/kB",`

I don't think a formatted string is a good way to represent these on the RPC. Better to not leave the clients to strip off the unit, leave this as number and report the unit in the documentation.

**Review by achow101 — APPROVED** [review]
utACK 983e4e9b005e54307f315f00f8ea90959482376c

**Comment by sipa** [comment]
Concept ACK

File: src/rpc/rawtransaction.cpp [hunk 2]
```diff
@@ -1862,23 +1862,23 @@ UniValue analyzepsbt(const JSONRPCRequest& request)
...
+                "      \"next\" : \"role\"           (string, optional) Role of the next person that this input needs to go to\n"
                 "    }\n"
                 "    ,...\n"
                 "  ]\n"
-                "  \"estimated_vsize\" : vsize       (numeric) Estimated vsize of the final signed transaction\n"
+                "  \"estimated_vsize\" : vsize       (numeric, optional) Estimated vsize of the final signed transaction\n"
                 "  \"estimated_feerate\" : feerate   (numeric, optional) Estimated feerate of the final signed transaction. Shown only if all UTXO slots in the PSBT have been filled.\n"
                 "  \"fee\" : fee                     (numeric, optional) The transaction fee paid. Shown only if all UTXO slots in the PSBT have been filled.\n"
-                "  \"next\" : \"role\"                 (string) Role of the next person that this psbt needs to go to\n"
+                "  \"next\" : \"role\"               (string) Role of the next person that this psbt needs to go to\n"
```
Comment (MarcoFalke) [inline src/rpc/rawtransaction.cpp]:
```suggestion
                "  \"next\" : \"role\"                 (string) Role of the next person that this psbt needs to go to\n"
```

File: src/rpc/rawtransaction.cpp [hunk 3]
```diff
@@ -1862,23 +1862,23 @@ UniValue analyzepsbt(const JSONRPCRequest& request)
...
+                "        \"signatures\" : [          (array, optional)\n"
                 "          \"keyid\"                 (string) Public key ID, hash160 of the public key, of a public key whose signature is missing\n"
                 "        ]\n"
-                "        \"redeemscript\" : \"hash\"   (string) Hash160 of the redeemScript that is missing\n"
-                "        \"witnessscript\" : \"hash\"  (string) SHA256 of the witnessScript that is missing\n"
+                "        \"redeemscript\" : \"hash\"   (string, optional) Hash160 of the redeemScript that is missing\n"
+                "        \"witnessscript\" : \"hash\"  (string, optional) SHA256 of the witnessScript that is missing\n"
                 "      }\n"
-                "      \"next\" : \"role\"           (string) Role of the next person that this input needs to go to\n"
+                "      \"next\" : \"role\"           (string, optional) Role of the next person that this input needs to go to\n"
```
Comment (MarcoFalke) [inline src/rpc/rawtransaction.cpp]:
```suggestion
                "      \"next\" : \"role\"             (string, optional) Role of the next person that this input needs to go to\n"
```

File: src/rpc/rawtransaction.cpp [hunk 4]
```diff
@@ -1862,23 +1862,23 @@ UniValue analyzepsbt(const JSONRPCRequest& request)
...
                 "    }\n"
                 "    ,...\n"
                 "  ]\n"
-                "  \"estimated_vsize\" : vsize       (numeric) Estimated vsize of the final signed transaction\n"
+                "  \"estimated_vsize\" : vsize       (numeric, optional) Estimated vsize of the final signed transaction\n"
                 "  \"estimated_feerate\" : feerate   (numeric, optional) Estimated feerate of the final signed transaction. Shown only if all UTXO slots in the PSBT have been filled.\n"
```
Comment (MarcoFalke) [inline src/rpc/rawtransaction.cpp]:
```suggestion
                "  \"estimated_feerate\" : feerate   (numeric, optional) Estimated feerate of the final signed transaction in " + CURRENCY_UNIT + "/kB. Shown only if all UTXO slots in the PSBT have been filled.\n"
```

**Review by maflcko — APPROVED** [review]
ACK

**Comment by fanquake** [comment]
Updated with marcos suggestions.

```src/bitcoin-cli analyzepsbt```
```
analyzepsbt "psbt"

Analyzes and provides information about the current status of a PSBT and its inputs

Arguments:
1. psbt    (string, required) A base64 string of a PSBT

Result:
{
  "inputs" : [                      (array of json objects)
    {
      "has_utxo" : true|false     (boolean) Whether a UTXO is provided
      "is_final" : true|false     (boolean) Whether the input is finalized
      "missing" : {               (json object, optional) Things that are missing that are required to complete this input
        "pubkeys" : [             (array, optional)
          "keyid"                 (string) Public key ID, hash160 of the public key, of a public key whose BIP 32 derivation path is missing
        ]
        "signatures" : [          (array, optional)
          "keyid"                 (string) Public key ID, hash160 of the public key, of a public key whose signature is missing
        ]
        "redeemscript" : "hash"   (string, optional) Hash160 of the redeemScript that is missing
        "witnessscript" : "hash"  (string, optional) SHA256 of the witnessScript that is missing
      }
      "next" : "role"             (string, optional) Role of the next person that this input needs to go to
    }
    ,...
  ]
  "estimated_vsize" : vsize       (numeric, optional) Estimated vsize of the final signed transaction
  "estimated_feerate" : feerate   (numeric, optional) Estimated feerate of the final signed transaction in BTC/kB. Shown only if all UTXO slots in the PSBT have been filled.
  "fee" : fee                     (numeric, optional) The transaction fee paid. Shown only if all UTXO slots in the PSBT have been filled.
  "next" : "role"                 (string) Role of the next person that this psbt needs to go to
}

Examples:
> bitcoin-cli analyzepsbt "psbt"
```

**Comment by MarcoFalke** [comment]
utACK 335931df4a37467299a2ee0ba521ecd5c4e6d39e

**Comment by laanwj** [comment]
utACK 335931df4a37467299a2ee0ba521ecd5c4e6d39e

**Comment by fanquake** [comment]
Has been backported in https://github.com/bitcoin/bitcoin/commit/20fd64fe10410eb467262a168c18079f5dae5d52 and https://github.com/bitcoin/bitcoin/commit/2edd0c40c34bb4d0132df31b110005b7fe944170.

**Pull request merged**